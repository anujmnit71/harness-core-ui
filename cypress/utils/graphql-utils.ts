export const hasOperationName: (req: any, operationName: string) => boolean = (req, operationName) => {
  const { body } = req
  // eslint-disable-next-line no-prototype-builtins
  return body.hasOwnProperty('operationName') && body.operationName === operationName
}

export const aliasQuery: (req: any, operationName: string) => void = (req, operationName) => {
  if (hasOperationName(req, operationName)) {
    req.alias = `gql${operationName}Query`
  }
}
