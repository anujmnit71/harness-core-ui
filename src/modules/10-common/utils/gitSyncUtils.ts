/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { GitSyncConfig } from 'services/cd-ng'
import { folderPathName, yamlFileExtension } from './StringUtils'

export const getRepoDetailsByIndentifier = (identifier: string | undefined, repos: GitSyncConfig[]) => {
  return repos.find((repo: GitSyncConfig) => repo.identifier === identifier)
}

export const validateFilePath = (filePath: string): boolean => {
  if (!filePath.endsWith(yamlFileExtension)) {
    return false
  } else {
    const fullPath = filePath.slice(0, filePath.length - yamlFileExtension.length)
    const subPaths = fullPath.split('/')
    return subPaths.every((folderName: string) => {
      return folderName.length && folderName.match(folderPathName)?.length
    })
  }
}
