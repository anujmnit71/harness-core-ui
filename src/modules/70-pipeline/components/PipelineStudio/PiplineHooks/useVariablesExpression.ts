/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { useEffect, useState } from 'react'
import { defaultTo, identity, isEmpty, map, sortBy, sortedUniq } from 'lodash-es'
import type { VariableMergeServiceResponse } from 'services/pipeline-ng'
import { useTemplateVariables } from '../../TemplateVariablesContext/TemplateVariablesContext'
import { usePipelineVariables } from '../../PipelineVariablesContext/PipelineVariablesContext'
import { usePipelineContext } from '../PipelineContext/PipelineContext'
/**
 * Traverse over stage and find out all local fqn
 */
function traverseStageObject(
  jsonObj: Record<string, any>,
  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
): string[] {
  const keys: string[] = []
  if (jsonObj !== null && typeof jsonObj == 'object') {
    Object.entries(jsonObj).forEach(([_key, value]) => {
      keys.push(...traverseStageObject(value, metadataMap))
    })
  } else if (metadataMap[jsonObj]) {
    keys.push(jsonObj)
  }
  return keys
}

/**
 * Hook to integrate and get expression for local stage and other stage
 */
export function useVariablesExpression(): { expressions: string[] } {
  const { variablesPipeline, metadataMap, initLoading } = usePipelineVariables()
  const { metadataMap: templateMetadataMap, initLoading: templateInitLoading } = useTemplateVariables()
  const [expressions, setExpressions] = useState<string[]>([])
  const [localStageKeys, setLocalStageKeys] = useState<string[]>([])
  const {
    state: { selectionState: { selectedStageId } = { selectedStageId: '' } },
    getStageFromPipeline
  } = usePipelineContext()

  useEffect(() => {
    if (!initLoading && selectedStageId && !isEmpty(selectedStageId)) {
      const stage = getStageFromPipeline(selectedStageId, variablesPipeline).stage
      if (stage) {
        const keys = traverseStageObject(stage, metadataMap)
        setLocalStageKeys(keys)
      }
    }
  }, [variablesPipeline, initLoading, selectedStageId, metadataMap, getStageFromPipeline])

  useEffect(() => {
    if (!initLoading && !isEmpty(metadataMap)) {
      const expression = sortedUniq(
        sortBy(
          map(metadataMap, (item, index) =>
            localStageKeys.indexOf(index) > -1
              ? defaultTo(item.yamlProperties?.localName, '')
              : defaultTo(item.yamlProperties?.fqn, '')
          ).filter(p => p),
          identity
        )
      )
      const outputExpression = sortedUniq(
        sortBy(
          map(metadataMap, (item, index) =>
            localStageKeys.indexOf(index) > -1
              ? defaultTo(item.yamlOutputProperties?.localName, '')
              : defaultTo(item.yamlOutputProperties?.fqn, '')
          ).filter(p => p),
          identity
        )
      )
      setExpressions([...expression, ...outputExpression])
    }
  }, [initLoading, metadataMap, localStageKeys])

  useEffect(() => {
    if (!templateInitLoading && !isEmpty(templateMetadataMap)) {
      const expression = sortedUniq(
        sortBy(
          map(templateMetadataMap, item => defaultTo(item.yamlProperties?.localName, '')).filter(p => p),
          identity
        )
      )
      const outputExpression = sortedUniq(
        sortBy(
          map(templateMetadataMap, item => defaultTo(item.yamlOutputProperties?.localName, '')).filter(p => p),
          identity
        )
      )
      setExpressions([...expression, ...outputExpression])
    }
  }, [templateInitLoading, templateMetadataMap])

  return { expressions }
}
