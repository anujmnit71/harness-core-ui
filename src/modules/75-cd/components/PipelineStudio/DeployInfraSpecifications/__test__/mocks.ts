/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const envs = {
  status: 'SUCCESS',
  data: {
    totalPages: 1,
    totalItems: 1,
    pageItemCount: 1,
    pageSize: 100,
    content: [
      {
        environment: {
          accountId: 'zEaak-FLS425IEO7OLzMUg',
          orgIdentifier: 'default',
          projectIdentifier: 'Test_Yunus',
          identifier: 'dev',
          name: 'something',
          description: '',
          color: '#0063F7',
          type: 'Production',
          deleted: false,
          tags: {},
          version: 2
        },
        createdAt: 1640688198443,
        lastModifiedAt: 1640699193440
      }
    ],
    pageIndex: 0,
    empty: false
  },
  metaData: null,
  correlationId: '99ca7534-0816-4723-a892-da2463604b8f'
}
