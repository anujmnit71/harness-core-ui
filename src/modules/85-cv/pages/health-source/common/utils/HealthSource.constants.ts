/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const HealthSourceFieldNames = {
  MANUAL_QUERY: 'isManualQuery',
  IDENTIFIER: 'identifier',
  METRIC_NAME: 'metricName',
  METRIC_DATA: 'metricData',
  GROUP_NAME: 'groupName',
  CONTINUOUS_VERIFICATION: 'continuousVerification',
  HEALTH_SCORE: 'healthScore',
  SLI: 'sli',
  RISK_CATEGORY: 'riskCategory',
  HIGHER_BASELINE_DEVIATION: 'higherBaselineDeviation',
  LOWER_BASELINE_DEVIATION: 'lowerBaselineDeviation'
}
