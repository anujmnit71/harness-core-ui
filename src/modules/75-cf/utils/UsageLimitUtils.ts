/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { ReactNode } from 'react'
import { BannerType } from '@common/layouts/Constants'
import type { CheckFeatureReturn } from 'framework/featureStore/featureStoreUtil'
import { formatToCompactNumber } from '@cf/utils/CFUtils'
import type { UseStringsReturn } from 'framework/strings'

export const getBannerText = (
  getString: UseStringsReturn['getString'],
  monthlyActiveUsers: CheckFeatureReturn,
  additionalLicenseProps: Record<string, any>
): ReactNode => {
  const { isEnterpriseEdition, isFreeEdition, isTeamEdition } = additionalLicenseProps

  const clientMauUsageCount = Number(monthlyActiveUsers?.featureDetail?.count)
  const clientMauPlanLimit = Number(monthlyActiveUsers?.featureDetail?.limit)

  const clientMauUsagePercentage = Math.trunc((clientMauUsageCount / clientMauPlanLimit) * 100)
  const clientMauPlanLimitFormatted = formatToCompactNumber(clientMauPlanLimit)

  const showInfoBanner = clientMauPlanLimit > 0 && clientMauUsagePercentage >= 90 && clientMauUsagePercentage < 100
  const showWarningBanner = clientMauPlanLimit > 0 && clientMauUsagePercentage >= 100

  let message!: string
  let bannerType!: BannerType

  if (isFreeEdition) {
    if (!showInfoBanner) {
      return {
        message: () =>
          getString('cf.planEnforcement.freePlan.approachingLimit', {
            clientMauUsagePercentage,
            clientMauPlanLimitFormatted
          }),
        bannerType: BannerType.INFO
      }
    }

    if (showWarningBanner) {
      return {
        message: getString('cf.planEnforcement.freePlan.upgradeRequired', {
          clientMauPlanLimitFormatted
        }),
        bannerType: BannerType.LEVEL_UP
      }
    }
  }

  if (isTeamEdition || isEnterpriseEdition) {
    if (showInfoBanner) {
      return {
        message: () =>
          getString('cf.planEnforcement.teamEnterprisePlan.approachingLimit', {
            clientMauUsagePercentage
          }),
        bannerType: BannerType.INFO
      }
    }

    if (showWarningBanner) {
      return {
        message: () => getString('cf.planEnforcement.teamEnterprisePlan.upgradeRequired'),
        bannerType: BannerType.LEVEL_UP
      }
    }
  }

  return {
    message,
    bannerType
  }
}
