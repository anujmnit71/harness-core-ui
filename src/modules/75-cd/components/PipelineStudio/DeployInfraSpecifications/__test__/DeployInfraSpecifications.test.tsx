/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, waitFor } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import {
  PipelineContextInterface,
  PipelineContext
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { StageType } from '@pipeline/utils/stageHelpers'
import { StageErrorContext } from '@pipeline/context/StageErrorContext'
import * as useValidationErrors from '@pipeline/components/PipelineStudio/PiplineHooks/useValidationErrors'
import { envs } from './mocks'
import overridePipelineContext from './overrideSetPipeline.json'
import DeployInfraSpecifications from '../DeployInfraSpecifications'

const getOverrideContextValue = (): PipelineContextInterface => {
  return {
    ...overridePipelineContext,
    getStageFromPipeline: jest.fn().mockReturnValue({
      stage: {
        stage: {
          name: 'Stage 3',
          identifier: 's3',
          type: StageType.DEPLOY,
          description: '',
          spec: {}
        }
      }
    }),
    updateStage: jest.fn(),
    updatePipeline: jest.fn()
  } as any
}

jest.mock('services/cd-ng', () => ({
  useGetEnvironmentList: jest.fn().mockImplementation(() => ({ loading: false, data: envs, refetch: jest.fn() }))
}))

jest.mock('lodash-es', () => ({
  ...(jest.requireActual('lodash-es') as Record<string, any>),
  debounce: jest.fn(fn => {
    fn.cancel = jest.fn()
    return fn
  })
}))

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const intersectionObserverMock = () => ({
  observe: () => null,
  unobserve: () => null
})

window.IntersectionObserver = jest.fn().mockImplementation(intersectionObserverMock)

jest.mock('@pipeline/components/AbstractSteps/StepWidget', () => ({
  ...(jest.requireActual('@pipeline/components/AbstractSteps/StepWidget') as any),
  // eslint-disable-next-line react/display-name
  StepWidget: (props: any) => {
    return (
      <div className="step-widget-mock">
        <button
          name={'updateStepWidget'}
          onClick={() => {
            props.onUpdate({
              environment: {},
              environmentRef: 'dev'
            })
          }}
        >
          Step Widget button
        </button>
      </div>
    )
  }
}))

jest.mock(
  '@cd/components/PipelineStudio/DeployInfraSpecifications/SelectInfrastructureType/SelectInfrastructureType',
  // eslint-disable-next-line react/display-name
  () => (props: any) =>
    (
      <div className="select-infrastucture-type-mock">
        <button
          name={'updateSelectInfrastructureType'}
          onClick={() => {
            props.onChange('KubernetesDirect')
          }}
        >
          Select Infrastructure Type button
        </button>
      </div>
    )
)

describe('Deploy infra specifications test', () => {
  test('Should match snapshot', () => {
    const context = getOverrideContextValue()
    const { container } = render(
      <TestWrapper>
        <PipelineContext.Provider value={context}>
          <DeployInfraSpecifications />
        </PipelineContext.Provider>
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test(`Should Environment section be present`, () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={getOverrideContextValue()}>
          <DeployInfraSpecifications />
        </PipelineContext.Provider>
      </TestWrapper>
    )

    expect(getByText('environment')).toBeTruthy()
  })

  test(`Should Infrastructure definition section be present`, () => {
    const { getByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={getOverrideContextValue()}>
          <DeployInfraSpecifications />
        </PipelineContext.Provider>
      </TestWrapper>
    )

    expect(getByText('pipelineSteps.deploy.infrastructure.infraDefinition')).toBeTruthy()
  })

  test('Should call submitFormsForTab when errorMap is not empty', async () => {
    const errorContextProvider = {
      state: {} as any,
      checkErrorsForTab: jest.fn().mockResolvedValue(Promise.resolve()),
      subscribeForm: () => undefined,
      unSubscribeForm: () => undefined,
      submitFormsForTab: jest.fn()
    }

    jest.spyOn(useValidationErrors, 'useValidationErrors').mockReturnValue({ errorMap: new Map([['error', []]]) })

    render(
      <TestWrapper>
        <PipelineContext.Provider value={getOverrideContextValue()}>
          <StageErrorContext.Provider value={errorContextProvider}>
            <DeployInfraSpecifications />
          </StageErrorContext.Provider>
        </PipelineContext.Provider>
      </TestWrapper>
    )

    expect(errorContextProvider.submitFormsForTab).toBeCalled()
  })

  test(`Should updateEnvStep be called upon StepWidget change`, async () => {
    const context = getOverrideContextValue()
    const { findByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={context}>
          <DeployInfraSpecifications />
        </PipelineContext.Provider>
      </TestWrapper>
    )

    const button = await waitFor(() => findByText('Step Widget button'))
    fireEvent.click(button)

    expect(context.updateStage).toBeCalled()
  })

  test('Should Accordion be shown when deployment type is selected', async () => {
    const context = getOverrideContextValue()
    const { container, findByText } = render(
      <TestWrapper>
        <PipelineContext.Provider value={context}>
          <DeployInfraSpecifications />
        </PipelineContext.Provider>
      </TestWrapper>
    )

    const button = await waitFor(() => findByText('Select Infrastructure Type button'))

    act(() => {
      fireEvent.click(button)
    })

    expect(container.getElementsByClassName('accordion')[0]).toBeInTheDocument()
    expect(container.getElementsByClassName('tabHeading')[0]).toBeInTheDocument()
  })
})
