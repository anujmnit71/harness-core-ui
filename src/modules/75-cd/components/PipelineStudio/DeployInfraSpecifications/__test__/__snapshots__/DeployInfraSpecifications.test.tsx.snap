// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deploy infra specifications test Should match snapshot 1`] = `
<div>
  <div
    class="serviceOverrides"
  >
    <div
      class="contentSection"
    >
      <div
        class="tabHeading"
        id="environment"
      >
        environment
      </div>
      <div
        class="bp3-card bp3-elevation-0 Card--card sectionCard"
      >
        <div
          class="step-widget-mock"
        >
          <button
            name="updateStepWidget"
          >
            Step Widget button
          </button>
        </div>
      </div>
      <div
        class="tabHeading"
        id="infrastructureDefinition"
      >
        <span
          class="align"
          data-tooltip-id="pipelineStep.infrastructureDefinition"
        >
          pipelineSteps.deploy.infrastructure.infraDefinition
        </span>
      </div>
      <div
        class="bp3-card bp3-elevation-0 Card--card sectionCard"
      >
        <p
          class="StyledProps--font StyledProps--main info StyledProps--margin-bottom-medium"
        >
          <span
            class="align"
            data-tooltip-id="pipelineStep.infrastructureDefinitionMethod"
          >
            pipelineSteps.deploy.infrastructure.selectMethod
          </span>
        </p>
        <div
          class="select-infrastucture-type-mock"
        >
          <button
            name="updateSelectInfrastructureType"
          >
            Select Infrastructure Type button
          </button>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main StyledProps--margin-top-xxlarge"
      />
    </div>
  </div>
</div>
`;
